### PR TITLE
Refactor vof curvature l2 projection

### PR DESCRIPTION
**Pull request by AmishgaAlphonius**
### Description

Following PR #1318, this one refactors the curvature projection to be assembled and solved through the VOF subequations interface. Assembly of the matrix and right-hand side for both VOFSubequations (phase fraction gradient and curvature projection) are now done in the same loop without the WorkStream method, as they are pretty simple to assemble.

### Testing

The unit test dedicated to the interface has been updated.
Visual inspection of data was done with the rising bubble case using paraview.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] The branch is rebased onto master
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [x] If parameters are modified, the tests and the documentation of examples are up to date
- [x] Changelog (CHANGELOG.md) is up to date if the refactor affects the user experience or the codebase

Pull request related list:
- [x] No other PR is open related to this refactoring
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If any future works is planed, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: include/solvers/vof_subequations_interface.h [hunk 1]
```diff
@@ -11,15 +11,19 @@
 
 using namespace dealii;
 
+
 /**
  * @brief IDs associated to the different subequations solved in Lethe.
  */
 enum VOFSubequationsID : unsigned int
```
Comment (blaisb) [inline include/solvers/vof_subequations_interface.h:18]: can you make this an enum class or it needs to be an enum?

File: include/solvers/vof_subequations_interface.h [hunk 2]
```diff
@@ -284,6 +197,8 @@ class VOFSubequationsInterface
 
     if (subequation_id == VOFSubequationsID::phase_gradient_projection)
       subequation_string = "VOF phase fraction gradient L2 projection";
+    else if (subequation_id == VOFSubequationsID::curvature_projection)
+      subequation_string = "VOF curvature L2 projection";
```
Comment (blaisb) [inline include/solvers/vof_subequations_interface.h:201]: what is the subequation string used for?

File: source/solvers/vof.cc [hunk 3]
```diff
@@ -1539,194 +1538,6 @@ VolumeOfFluid<dim>::solve_projection_phase_fraction(GlobalVectorType &solution)
   solution = completely_distributed_phase_fraction_solution;
 }
 
-template <int dim>
```
Comment (blaisb) [inline source/solvers/vof.cc:1542]: Nice cleanup this is great.

File: source/solvers/vof_curvature_projection.cc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  std::vector<Tensor<1, dim>> grad_phi(n_dofs_per_cell);
+
+  // Get the diffusion factor
+  const double diffusion_factor =
+    this->simulation_parameters.multiphysics.vof_parameters
+      .surface_tension_force.curvature_diffusion_factor;
+
+  // Cell size
+  double h;
+
```
Comment (blaisb) [inline source/solvers/vof_curvature_projection.cc:65]: No need to pre-declare it, declare it when you get it.

File: source/solvers/vof_curvature_projection.cc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            phase_gradient_projection_cell);
+
+          // Get vector of Jacobi determinant times the quadrature weights
+          std::vector<double> JxW_vec =
+            fe_values_curvature_projection.get_JxW_values();
+
+          // Compute cell size
+          auto &fe_curvature_projection =
+            fe_values_curvature_projection.get_fe();
+          h = compute_cell_diameter<dim>(compute_cell_measure_with_JxW(JxW_vec),
```
Comment (blaisb) [inline source/solvers/vof_curvature_projection.cc:94]:
```suggestion
          const double h = compute_cell_diameter<dim>(compute_cell_measure_with_JxW(JxW_vec),
```

File: source/solvers/vof_curvature_projection.cc [hunk 6]
```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: Copyright (c) 2020-2024 The Lethe Authors
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
+#include <solvers/vof_curvature_projection.h>
+
+template <int dim>
+void
+VOFCurvatureProjection<dim>::assemble_system_matrix_and_rhs()
```
Comment (blaisb) [inline source/solvers/vof_curvature_projection.cc:8]: This is nice and easy to read, good job

File: source/solvers/vof_phase_gradient_projection.cc [hunk 7]
```diff
@@ -3,67 +3,154 @@
...
+  std::vector<Tensor<2, dim>> grad_phi(n_dofs_per_cell);
+
+  // Get the diffusion factor
+  const double diffusion_factor =
+    this->simulation_parameters.multiphysics.vof_parameters
+      .surface_tension_force.phase_fraction_gradient_diffusion_factor;
+
+  // Cell size
+  double h;
+
```
Comment (blaisb) [inline source/solvers/vof_phase_gradient_projection.cc:63]: Same here, declare h when you need it.

File: source/solvers/vof_phase_gradient_projection.cc [hunk 8]
```diff
@@ -3,67 +3,154 @@
...
 #include <solvers/vof_phase_gradient_projection.h>
 
-template <int dim, typename ScratchDataType>
+template <int dim>
 void
-VOFPhaseGradientProjection<dim, ScratchDataType>::assemble_local_system_matrix(
-  const typename DoFHandler<dim>::active_cell_iterator &cell,
-  ScratchDataType                                      &scratch_data,
-  StabilizedMethodsCopyData                            &copy_data)
+VOFPhaseGradientProjection<dim>::assemble_system_matrix_and_rhs()
```
Comment (blaisb) [inline source/solvers/vof_phase_gradient_projection.cc:8]: This is very easy to read, thanks

File: source/solvers/vof_phase_gradient_projection.cc [hunk 9]
```diff
@@ -3,67 +3,154 @@
...
+          fe_values_vof.reinit(vof_cell);
+
+          // Get vector of Jacobi determinant times the quadrature weights
+          std::vector<double> JxW_vec =
+            fe_values_phase_gradient_projection.get_JxW_values();
+
+          // Compute cell size
+          auto &fe_phase_gradient_projection =
+            fe_values_phase_gradient_projection.get_fe();
+          h = compute_cell_diameter<dim>(compute_cell_measure_with_JxW(JxW_vec),
```
Comment (blaisb) [inline source/solvers/vof_phase_gradient_projection.cc:91]:
```suggestion
          const double h = compute_cell_diameter<dim>(compute_cell_measure_with_JxW(JxW_vec),
```

**Review by blaisb — CHANGES_REQUESTED** [review]
Last few changes, quasi nothing. Very good work.

**Review by oguevremont — APPROVED** [review]
Looks good!

**Review by hepap — APPROVED** [review]
Not a lot of comments! I like this new interface a lot!

**Review by blaisb — APPROVED** [review]

**Pull request merged**